### ISSUE TITLE
Regression: Animation does not play when playAnimation() called in RecyclerView onBindViewHolder

### ISSUE THREAD

**Issue by kcoppock** [airbnb/lottie-android#1243]
When setting an animation from a `RecyclerView.Adapter`'s `onBindViewHolder()` method, the animation will never play. This is a regression from 3.0.0 to 3.0.1, and I believe is caused by commit a483b7695e.

Essentially, when `onBindViewHolder()` is first called, the view is not attached to the window yet. With this change, the animation will not start but will set the `wasAnimatingWhenNotShown` flag to true. However, in `onAttachedToWindow()`, both the `autoPlay` and `wasAnimatingWhenNotShown` flags are checked to decide whether or not to play the animation. If the `lottie_autoPlay` attribute was not set in XML, then the animation will never start playback unless the `playAnimation()` call is deferred to after `onAttachedToWindow()`. 

I've attached a simple sample project here ([LottieBug.zip](https://github.com/airbnb/lottie-android/files/3237744/LottieBug.zip)) that reproduces the issue, but it can be reproduced as simply as the following:

```
class MainActivity : AppCompatActivity() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    
    setContentView(
      RecyclerView(this).apply {
        layoutManager = LinearLayoutManager(this@MainActivity, LinearLayoutManager.VERTICAL, false)
        adapter = object : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
          override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            return object : RecyclerView.ViewHolder(LottieAnimationView(parent.context)) {}
          }

          override fun getItemCount(): Int = 1

          override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            (holder.itemView as LottieAnimationView).apply {
              loop(true)
              setAnimation("lottie.json")
              playAnimation()
            }
          }
        }
      }
    )
  }
}
```

**Comment by g-coppock** [comment]
A workaround is to either set `lottie_autoPlay` in XML, if possible, or the following:

```
if (ViewCompat.isAttachedToWindow(this)) {
  playAnimation()
} else {
  post(::playAnimation)
}
```

**Comment by kcoppock** [comment]
A workaround is to either set `lottie_autoPlay` in XML, if possible, or the following:

```
if (ViewCompat.isAttachedToWindow(this)) {
  playAnimation()
} else {
  post(::playAnimation)
}
```

**Comment by gpeal** [comment]
@kcoppock Thanks for the repro steps. Looking into it now.

**Comment by gpeal** [comment]
@kcoppock Hmm, in both your sample project and and the code snipped you provided, the animation plays on launch which is the expected behavior right?

**Comment by gpeal** [comment]
@kcoppock That being said, upon first glance, it looks like the correct version of
```
    if (autoPlay && wasAnimatingWhenDetached) {
      playAnimation();
    }
```
would be
```
    if (autoPlay || wasAnimatingWhenDetached) {
      playAnimation();
      autoPlay = false;
    }
```

That would ensure that the attr only takes effect once and `wasAnimatingWhenDetached` will always have an effect.

**Comment by kcoppock** [comment]
Ah sorry about that, this seems like it may be partially platform-specific. I can easily repro with an API 21 emulator, but not with an API 28 Pixel 3. However, even with the Pixel 3 you can reproduce the issue if you add a bunch of items the adapter. For example, change the `listOf("lottie.json")` to `(0 until 100).map { "lottie.json" }`. 

Scroll up just until one of the views goes offscreen and then scroll it back down into view. This should trigger it to be detached from the window and then reattached, and you'll see the issue reproduce.

Your fix does sound reasonable to me though, if autoPlay is only intended to take effect once.

**Comment by gpeal** [comment]
@kcoppock Alright, I can repro it and it seems to be happening because onVisibilityChanged isn't called on older API levels.... I'm going to look into ways to get around this

**Comment by kcoppock** [comment]
Great, thanks! Let me know if I can help. This looks relevant:

https://stackoverflow.com/questions/53237535/os-version-4-4-kitkat-and-5-lollipop-onvisibilitychanged-listener-not-firing

**Comment by gpeal** [comment]
@kcoppock Yup, that looks spot on!
